### ISSUE TITLE
Crash after inserting >100 keys

### ISSUE THREAD

**Issue by tv42** [kellydunn/go-art#2]
```
package main

import (
	"encoding/binary"
	"math/rand"

	"github.com/kellydunn/go-art"
)

func main() {
	rand.Seed(42)
	key := make([]byte, 8)
	tree := art.NewArtTree()
	for i := 0; i < 135; i++ {
		binary.BigEndian.PutUint64(key, uint64(rand.Int63()))
		tree.Insert(key, key)
	}
}
```

```
$ go run bug.go 
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/kellydunn/go-art.(*ArtTree).insertHelper(0xc20800a1f0, 0xc20803f170, 0xc20805c3b0, 0xc20800aca0, 0x9, 0x10, 0x483380, 0xc20801f5c0, 0x1)
	/home/tv/go/src/github.com/kellydunn/go-art/art_tree.go:115 +0x619
github.com/kellydunn/go-art.(*ArtTree).insertHelper(0xc20800a1f0, 0xc20803e090, 0xc20800a1f0, 0xc20800aca0, 0x9, 0x10, 0x483380, 0xc20801f5c0, 0x0)
	/home/tv/go/src/github.com/kellydunn/go-art/art_tree.go:170 +0xde7
github.com/kellydunn/go-art.(*ArtTree).Insert(0xc20800a1f0, 0xc20800aca0, 0x9, 0x10, 0x483380, 0xc20801f5c0)
	/home/tv/go/src/github.com/kellydunn/go-art/art_tree.go:62 +0xa9
main.main()
	/home/tv/tmp/max.go:16 +0x154

goroutine 2 [runnable]:
runtime.forcegchelper()
	/home/tv/src/go1.4/src/runtime/proc.go:90
runtime.goexit()
	/home/tv/src/go1.4/src/runtime/asm_amd64.s:2232 +0x1

goroutine 3 [runnable]:
runtime.bgsweep()
	/home/tv/src/go1.4/src/runtime/mgc0.go:82
runtime.goexit()
	/home/tv/src/go1.4/src/runtime/asm_amd64.s:2232 +0x1

goroutine 4 [runnable]:
runtime.runfinq()
	/home/tv/src/go1.4/src/runtime/malloc.go:712
runtime.goexit()
	/home/tv/src/go1.4/src/runtime/asm_amd64.s:2232 +0x1
exit status 2
```

**Comment by kellydunn** [comment]
Interesting find!

I'm not sure if its because you're trying to insert more than 100 keys, but it looks like it's currently blowing up on this line: https://github.com/kellydunn/go-art/blob/master/art_tree.go#L115-L116

Upon first glance, it looks like the lib might be making assumptions about path compression.

I'll take a deeper look when I get home!  Thanks for filling the issue :)

**Comment by kellydunn** [comment]
Update: 

It looks like the root cause of the bug is in the slicing / array copy semantics in the insert logic.  When I increased the size of the key in your example, this program exited successfully:

```golang
package main

import (
	"encoding/binary"
	"fmt"
	"github.com/kellydunn/go-art"
	"math/rand"
)

func main() {
	rand.Seed(42)
	key := make([]byte, 16)
	tree := art.NewArtTree()
	for i := 0; i < 135; i++ {
		binary.BigEndian.PutUint64(key, uint64(rand.Int63()))
		fmt.Printf("%v\n", key)
		tree.Insert(key, key)
	}
}
```

This is because there are some assumptions in the code that the passed in `[]byte` will be a byte slice and not a byte array with a set capacity.  I'll work on fixing it!

**Comment by tv42** [comment]
Behind every (non-nil) slice is an array. Every slice has a set capacity. This sounds weird.

**Comment by kellydunn** [comment]
Sorry, you are correct; I was mistaken. 

What I was trying to articulate was that there's an index out of bounds error during path compression calculations.  This is because `go-art` finds that the desired insertion key matches the entire child node at a certain depth:

```
Depth: 1. Longest Common Prefix: 7 bytes.  Current Key: [118 182 13 246 62 61 236 66].  Other Key: [118 182 13 246 62 61 236 66 0].
panic: runtime error: index out of range
```

It's at this point that `go-art` wants to insert the path-compressed key `182 13 246 62 61 236 66` into the tree, but cannot because it assumes that there's a null terminated byte at the end of this key, and thusly indexes out of its bounds.  I should add that before insertion, `go-art` null-terminates keys due to leaf nodes being incorrectly inserted when similar keys will want to be inserted at the same internal node (example: "apple" and "apple sauce")

For some reason, the current key in the iteration is not null-terminated for this particular node.  I'll do some further investigation to find out why.

**Comment by tv42** [comment]
Are you relying on keys to not contain the zero byte? That'd be.. less than ideal.

**Comment by kellydunn** [comment]
The user can enter whatever they want; the zero byte is more of an internal implementation detail.  It just looks like it's misbehaving in this instance.

**Comment by kellydunn** [comment]
@tv42 after a bit more research, it looks like the underlying `[]byte` that were storing keys in the tree were addressing the same memory as the `[]byte` your code example was using to insert into the tree.

I have created a Pull request #3 , which should remedy this issue.  It explicitly copies the contents of the passed in key value before creating the new Node and adding it to the tree.

**Comment by kellydunn** [comment]
Closing, as this should be resolved in #3
